### PR TITLE
#43 [Madhavan|Jai] Integrating Snyk Scan to CI

### PR DESCRIPTION
**Pull request by jai-charan**
- Added Snyc scan to run on the CI pipeline
- Runs everytime the pipeline is triggered and generates the report of Vulnerablities

### CODE REVIEW

File: .github/workflows/build-and-test.yml [hunk 1]
```diff
@@ -30,7 +26,21 @@ jobs:
 
       - name: Build all on ${{matrix.os}}
         run: make all
-        
+
```
Comment (vraravam) [inline .github/workflows/build-and-test.yml:27]: pls remove empty lines

**Review by vraravam — CHANGES_REQUESTED** [review]

File: .github/workflows/build-and-test.yml [hunk 2]
```diff
@@ -30,7 +26,21 @@ jobs:
...
+      - name: Run Snyk security scan
+        run: |
+          snyk auth ${{ secrets.SNYK_TOKEN }}
+          snyk test --all-projects
+
+      - name: Run Snyk Monitor
+        run: |
+          snapshot_link=$(snyk monitor --all-projects | grep "Explore this snapshot at" | sed 's/Explore this snapshot at //')
+          echo "Snyk Monitor link: $snapshot_link" >> $GITHUB_STEP_SUMMARY
```
Comment (vraravam) [inline .github/workflows/build-and-test.yml]: Make this to be markdown syntax for cleaner visible text

Comment (vraravam) [inline .github/workflows/build-and-test.yml]:
```suggestion
          echo "[Snyk Monitor link]($snapshot_link)" >> $GITHUB_STEP_SUMMARY
```

**Review by vraravam — CHANGES_REQUESTED** [review]

File: .github/workflows/build-and-test.yml (hunk 1 above)
Comment (jai-charan) [reply .github/workflows/build-and-test.yml:27]: Okay

File: .github/workflows/build-and-test.yml (hunk 2 above)
Comment (jai-charan) [reply .github/workflows/build-and-test.yml]: okayy

Comment (jai-charan) [reply .github/workflows/build-and-test.yml]: okay vijay

File: .github/workflows/build-and-test.yml [hunk 3]
```diff
@@ -30,7 +26,21 @@ jobs:
 
       - name: Build all on ${{matrix.os}}
         run: make all
-        
+
+
+      - name: Install Snyk CLI
+        run: npm install -g snyk
```
Comment (vraravam) [inline .github/workflows/build-and-test.yml]: Better to always have a locked/known version

File: .github/workflows/build-and-test.yml [hunk 4]
```diff
@@ -1,21 +1,17 @@
...
     branches: [ master ]
-
 jobs:
-
   build_and_test:
     name: Build and test on ${{ matrix.os }}
     runs-on: ${{ matrix.os }}
     strategy:
       matrix:
         os: [ubuntu-latest]
```
Comment (vraravam) [inline .github/workflows/build-and-test.yml]: Better to always have a locked/known version

File: .github/workflows/build-and-test.yml [hunk 5]
```diff
@@ -1,21 +1,17 @@
...
     branches: [ master ]
   pull_request:
     branches: [ master ]
-
 jobs:
-
   build_and_test:
     name: Build and test on ${{ matrix.os }}
     runs-on: ${{ matrix.os }}
     strategy:
```
Comment (vraravam) [inline .github/workflows/build-and-test.yml]: since you are not going to use different OSes, you can simplify this portion of the GHA workflow

**Review by vraravam — CHANGES_REQUESTED** [review]

File: .github/workflows/build-and-test.yml (hunk 5 above)
Comment (jai-charan) [reply .github/workflows/build-and-test.yml]: Made changes as mentioned.

File: .github/workflows/build-and-test.yml [hunk 6]
```diff
@@ -1,21 +1,15 @@
...
-
   build_and_test:
-    name: Build and test on ${{ matrix.os }}
-    runs-on: ${{ matrix.os }}
-    strategy:
-      matrix:
-        os: [ubuntu-latest]
-    steps:
+    name: Build and test on ubuntu-latest
+    runs-on: ubuntu-latest
```
Comment (vraravam) [inline .github/workflows/build-and-test.yml]: dont use `ubuntu-latest` - use a specific version

File: .github/workflows/build-and-test.yml [hunk 7]
```diff
@@ -28,9 +22,24 @@ jobs:
       - name: Check out code into the Go module directory
         uses: actions/checkout@v2
 
-      - name: Build all on ${{matrix.os}}
+      - name: Build all on ubuntu-latest
```
Comment (vraravam) [inline .github/workflows/build-and-test.yml]: don't use `ubuntu-latest`; instead, use a specific version

**Review by vraravam — CHANGES_REQUESTED** [review]

File: .github/workflows/build-and-test.yml (hunk 3 above)
Comment (jai-charan) [reply .github/workflows/build-and-test.yml]: Made changes

File: .github/workflows/build-and-test.yml (hunk 4 above)
Comment (jai-charan) [reply .github/workflows/build-and-test.yml]: Locked on ubuntu-20.04

File: .github/workflows/build-and-test.yml (hunk 6 above)
Comment (jai-charan) [reply .github/workflows/build-and-test.yml]: Locked on ubuntu-20.04

File: .github/workflows/build-and-test.yml [hunk 8]
```diff
@@ -39,7 +48,7 @@ jobs:
 
   publish:
```
Comment (vraravam) [inline .github/workflows/build-and-test.yml]: why do we need a separate stage called `publish`? Why can't the last step be merged into the above stage itself?

Comment (jai-charan) [reply .github/workflows/build-and-test.yml]: Merged publish stage to the existing stage

File: .github/workflows/build-and-test.yml [hunk 9]
```diff
@@ -28,22 +22,30 @@ jobs:
...
+        run: |
+          snyk auth ${{ secrets.SNYK_TOKEN }}
+          snyk test --all-projects
+
+      - name: Run Snyk Monitor
+        run: |
+          snapshot_link=$(snyk monitor --all-projects | grep "Explore this snapshot at" | sed 's/Explore this snapshot at //')
+          echo "[Snyk Monitor link]($snapshot_link)" >> $GITHUB_STEP_SUMMARY
+
       - name: Upload test report
```
Comment (vraravam) [inline .github/workflows/build-and-test.yml]: now that we have merged the `publish` step, do we still need to uplaod and download the test report?

**Review by vraravam — CHANGES_REQUESTED** [review]

**Review by vraravam — APPROVED** [review]

**Review by vraravam — APPROVED** [review]

**Pull request merged**